diff -ruN init.samsung/init.rc init.b2g/init.rc
--- init.samsung/init.rc	2012-07-08 15:44:20.781792634 +0300
+++ init.b2g/init.rc	2012-05-21 15:33:50.000000000 +0300
@@ -15,7 +15,9 @@
 
 # setup the global environment
     export PATH /sbin:/vendor/bin:/system/sbin:/system/bin:/system/xbin
-    export LD_LIBRARY_PATH /vendor/lib:/system/lib
+    export TMPDIR /data/local/tmp
+    export GRE_HOME /system/b2g
+    export LD_LIBRARY_PATH /vendor/lib:/system/lib:/system/b2g
     export ANDROID_BOOTLOGO 1
     export ANDROID_ROOT /system
     export ANDROID_ASSETS /system/app
@@ -106,7 +108,9 @@
 on fs
 # mount ext4 partitions
     # Mount /system rw first to give the filesystem a chance to save a checkpoint
-    mount ext4 /res/dev/system /system noatime wait ro
+    mount ext4 /res/dev/system /system noatime wait rw
+    chmod 0755 /system/b2g/b2g
+    chmod 0755 /system/b2g/updater
     
     exec sfsck /res/dev/cache ext4
     mount ext4 /res/dev/cache /cache nosuid nodev noatime wait
@@ -594,6 +598,7 @@
 
     class_start core
     class_start main
+    class_start b2g
 
 on nonencrypted
     class_start late_start
@@ -708,8 +713,6 @@
     user system
     group system
     critical
-    onrestart restart zygote
-    onrestart restart media
 
 service vold /system/bin/vold
     class core
@@ -727,7 +730,7 @@
 
 service ril-daemon /system/bin/rild
     class main
-    socket rild stream 660 root radio
+    socket rild stream 666 root radio
     socket rild-debug stream 660 radio system
     user root
     group radio cache inet misc audio sdcard_rw log
@@ -747,6 +750,7 @@
     user system
     group graphics
     onrestart restart zygote
+    disabled
 
 service zygote /system/bin/app_process -Xzygote /system/bin --zygote --start-system-server
     class main
@@ -755,6 +759,7 @@
     onrestart write /sys/power/state on
     onrestart restart media
     onrestart restart netd
+    disabled
 
 service drm /system/bin/drmserver
     class main
@@ -899,6 +904,26 @@
 #    class main
 #    user root
 #    oneshot
+
+service fakeperm /system/bin/fakeperm
+    class b2g
+    user root
+
+service b2g /system/b2g/b2g
+    class b2g
+    user root
+    onrestart restart media
+
+service rilproxy /system/bin/rilproxy
+    class b2g
+    socket rilproxy stream 660 root system
+    socket rilproxyd stream 660 root system
+    user root
+    group radio
+
+service mount-sdcard /system/bin/mountvol.sh sdcard
+    class b2g
+    oneshot
     
 # 2011-12-08/systemsw/kyo/ bugreport is triggered by holding down volume down, volume up and power
 service bugreport /system/bin/bugmailer.sh -v
diff -ruN init.samsung/init.rc.2 init.b2g/init.rc.2
--- init.samsung/init.rc.2	2012-05-21 15:35:42.000000000 +0300
+++ init.b2g/init.rc.2	2012-05-21 15:36:03.000000000 +0300
@@ -15,7 +15,9 @@
 
 # setup the global environment
     export PATH /sbin:/vendor/bin:/system/sbin:/system/bin:/system/xbin
-    export LD_LIBRARY_PATH /vendor/lib:/system/lib
+    export TMPDIR /data/local/tmp
+    export GRE_HOME /system/b2g
+    export LD_LIBRARY_PATH /vendor/lib:/system/lib:/system/b2g
     export ANDROID_BOOTLOGO 1
     export ANDROID_ROOT /system
     export ANDROID_ASSETS /system/app
@@ -106,7 +108,9 @@
 on fs
 # mount ext4 partitions
     # Mount /system rw first to give the filesystem a chance to save a checkpoint
-    mount ext4 /res/dev/system /system noatime wait ro
+    mount ext4 /res/dev/system /system noatime wait rw
+    chmod 0755 /system/b2g/b2g
+    chmod 0755 /system/b2g/updater
 
     # verify filesystem (UMS)
     exec sformat /dev/block/mmcblk0p11 vfat
@@ -596,6 +600,7 @@
 
     class_start core
     class_start main
+    class_start b2g
 
 on nonencrypted
     class_start late_start
@@ -710,8 +715,6 @@
     user system
     group system
     critical
-    onrestart restart zygote
-    onrestart restart media
 
 service vold /system/bin/vold
     class core
@@ -729,7 +732,7 @@
 
 service ril-daemon /system/bin/rild
     class main
-    socket rild stream 660 root radio
+    socket rild stream 666 root radio
     socket rild-debug stream 660 radio system
     user root
     group radio cache inet misc audio sdcard_rw log
@@ -749,6 +752,7 @@
     user system
     group graphics
     onrestart restart zygote
+    disabled
 
 service zygote /system/bin/app_process -Xzygote /system/bin --zygote --start-system-server
     class main
@@ -757,6 +761,7 @@
     onrestart write /sys/power/state on
     onrestart restart media
     onrestart restart netd
+    disabled
 
 service drm /system/bin/drmserver
     class main
@@ -901,6 +906,26 @@
 #    class main
 #    user root
 #    oneshot
+
+service fakeperm /system/bin/fakeperm
+    class b2g
+    user root
+
+service b2g /system/b2g/b2g
+    class b2g
+    user root
+    onrestart restart media
+
+service rilproxy /system/bin/rilproxy
+    class b2g
+    socket rilproxy stream 660 root system
+    socket rilproxyd stream 660 root system
+    user root
+    group radio
+
+service mount-sdcard /system/bin/mountvol.sh sdcard
+    class b2g
+    oneshot
     
 # 2011-12-08/systemsw/kyo/ bugreport is triggered by holding down volume down, volume up and power
 service bugreport /system/bin/bugmailer.sh -v
diff -ruN init.samsung/recovery.rc init.b2g/recovery.rc
--- init.samsung/recovery.rc	2012-07-08 15:46:18.000000000 +0300
+++ init.b2g/recovery.rc	2012-05-21 15:21:24.821842888 +0300
@@ -42,8 +42,8 @@
 
 service recovery /sbin/recovery-loader
 
-service adbd /sbin/adbd recovery
-    disabled
+service adbd /sbin/adbd
+    enabled
 
 # Always start adbd on userdebug and eng builds
 on property:ro.debuggable=1
@@ -52,8 +52,8 @@
 on property:adb.recovery=1
     write /sys/class/android_usb/android0/enable 0
     write /sys/class/android_usb/android0/idVendor 04e8
-    write /sys/class/android_usb/android0/idProduct 685e
-    write /sys/class/android_usb/android0/functions mass_storage,adb
+    write /sys/class/android_usb/android0/idProduct 6860
+    write /sys/class/android_usb/android0/functions adb
     write /sys/class/android_usb/android0/enable 1
     write /sys/class/android_usb/android0/iManufacturer $ro.product.manufacturer
     write /sys/class/android_usb/android0/iProduct $ro.product.model
